strawberry/permission.py

```diff
@@ -1,18 +1,46 @@
-from typing import Any, Awaitable, Optional, Union
+import abc
+from typing import TYPE_CHECKING, Any, Awaitable, Optional, Union
+from warnings import warn
 
+from strawberry.type import StrawberryOptional
 from strawberry.types.info import Info
 
 
-class BasePermission:
+if TYPE_CHECKING:
+    from strawberry.field import StrawberryField
+
+
+class BasePermission(abc.ABC):
     """
     Base class for creating permissions
     """
 
     message: Optional[str] = None
 
+    @abc.abstractclassmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs
     ) -> Union[bool, Awaitable[bool]]:
-        raise NotImplementedError(
-            "Permission classes should override has_permission method"
-        )
+        ...
+
+    @classmethod
+    def assert_valid_for_field(cls, field: "StrawberryField") -> None:
+        # assert all abstact methods are implemented in a subclass
+        not_implemented_methods = getattr(cls, "__abstractmethods__", None)
+        if not_implemented_methods:
+            method_names = ", ".join(not_implemented_methods)
+            raise NotImplementedError(
+                f"Permission class {cls.__name__} "
+                f"should have the following methods implemented: {method_names}"
+            )
```

Comment:
@BryceBeagle Thanks for the feedback. This is also what I thought at first. But the ABC check is performed during instantiation. So if this check is skipped inside the validation method, the schema initialization will not fail. And instead, the user will get the error querying any field that uses that invalid Permission class:
```graphql
{
  "data": null,
  "errors": [
    {
      "message": "Can't instantiate abstract class IsAuthenticated with abstract method has_permission",
      ...
    }
  ]
}
```
It is similar to how it works now just a different error message.
One idea I had was to try and instantiate the permission class inside the validation method:
```python
@classmethod
def assert_valid_for_field(cls, field: "StrawberryField") -> None:
    # assert all abstact methods are implemented in a subclass
    cls()
```
But that seemed less explicit, and if there are ever any parameters needed for the `__init__` this approach might not work.

I'll split the PR but let's first discuss this here, ok?

Reply:
Oh good point. Can we use [`inspect.isabstract`](https://docs.python.org/3/library/inspect.html#inspect.isabstract) here instead of the manual check against `__abstractmethods__`?